Mlem/Views/Shared/Composer/PostEditorDetailView+Logic.swift

```diff
@@ -0,0 +1,85 @@
+//
+//  PostEditorDetailView+Logic.swift
+//  Mlem
+//
+//  Created by Sjmarf on 24/09/2023.
+//
+
+import Foundation
+import SwiftUI
+import PhotosUI
+
+extension PostDetailEditorView {
+    var isReadyToPost: Bool {
+        switch uploadProgress {
+        case .noImage, .uploaded:
+            return postTitle.trimmed.isNotEmpty
+        default:
+            return false
+        }
+    }
+    
+    var isValidURL: Bool {
+        guard postURL.lowercased().hasPrefix("http://") ||
+            postURL.lowercased().hasPrefix("https://") else {
+            return false // URL protocol is missing
+        }
+
+        guard URL(string: postURL) != nil else {
+            return false // Not Parsable
+        }
+        
+        return true
+    }
+    
+    func submitPost() async {
+        do {
+            guard postTitle.trimmed.isNotEmpty else {
+                errorDialogMessage = "You need to enter a title for your post."
+                isShowingErrorDialog = true
+                return
+            }
+            
+            guard postURL.lowercased().isEmpty || isValidURL else {
+                errorDialogMessage = "You seem to have entered an invalid URL, please check it again."
+                isShowingErrorDialog = true
+                return
+            }
+            
+            isSubmitting = true
+            
+            try await onSubmit()
+            
+        } catch {
+            isSubmitting = false
+            errorHandler.handle(error)
+        }
+    }
+    
+    func uploadImage() {
+        uploadProgress = .uploading(0)
+        Task {
+            do {
+                let data = try await imageSelection?.loadTransferable(type: Data.self)
+                
+                if let data = data {
+                    if let uiImage = UIImage(data: data) {
+                        self.uploadedImage = Image(uiImage: uiImage)
+                    }
+                    let res = try await apiClient.uploadImage(data, callback: { uploadProgress = .uploading($0) })
+                    if let firstFile = res.files.first {
+                        var components = URLComponents()
+                        components.scheme = try apiClient.session.instanceUrl.scheme
+                        components.host = try apiClient.session.instanceUrl.host
+                        components.path = "/pictrs/image/\(firstFile.file)"
+                        postURL = components.url?.absoluteString ?? ""
+                        uploadProgress = .uploaded
+                    }
+                }
+            } catch {
+                uploadProgress = .failed(error)
+                return
+            }
+        }
```

Comment:
All of this logic should live in an `ImageRepository`

Reply:
How would I go about doing this?

Here's the `PictrsModel` struct that I'm using to represent an image at the moment:

```swift
struct PictrsImageModel {
    enum UploadState {
        case uploading(progress: Double)
        case uploaded(file: PictrsFile?)
        case failed(Error)
    }
    var image: Image? // a SwiftUI Image
    var state: UploadState = .uploading(progress: 0)
}
```

With the existing repositories, each function of the repository returns a completely new Model. The issue with replicating this for `PictrsImageModel` is that within the code block that loads the image, several mutating changes need to be made to an _existing_ struct. I don't see how this could be accomplished within an async context. Were you thinking of making `PictrsImageModel` a class instead of a struct? If so, how would I observe changes made to it? It can't be an `ObservableObject` because it needs to be optional, and having it observable would make managing multiple images at once difficult if we ever need to do that.

Maybe an option would be to make a `PictrsImageTracker` ObservableObject that stores a list of `PictrsImageModel` structs. A function of the tracker would take in a `PictrsImageModel` and run the above code, finding and replacing the struct in its published list of `PictrsImageModel` instances via some equation method (maybe UUID).

Apologies if I'm missing something obvious. Any suggestions would be appreciated.

Reply:
Hmm, yes, I see the complexity. There's probably a nice way to do this all, but for the time being I'm content to keep it as-is.

I still don't love having the `apiClient` directly called in view logic, but this is a weird case.

Reply:
Quick update - I've managed to make an image repository sort-of work by passing an `onUpdate` closure into the `upload` method. The `upload` method will call `onUpdate` with the new value of the `imageModel` whenever it makes a change, which the View can use to update it's own copy of the `imageModel`. It's not a perfect solution, but better than it was before I think